From: NeilBrown <neilb@suse.com>
Date: Thu, 4 Aug 2016 09:12:28 +1000
Subject: [PATCH] Revert "SUNRPC: xs_sock_mark_closed() does not need to
 trigger socket autoclose"
Patch-mainline: Submitted, 04 Aug 2016 09:33:15 +1000 - linux-nfs@vger.kernel.org
References: bsc#990356

This reverts commit 4b0ab51db32eba0f48b7618254742f143364a28d.

This change causes 'rpc.nfsd' to hang for long time if rpcbind is not
available.
If
   --no-nfs-version 2 --no-nfs-version 3
is given, the delay is about 6.5 minutes.  When trying to register
all versions, the delay is over half an hour.
Before this commit, and after reverting it, nfsd fails (when v3 is
requested) or succeeds (when only v4 is requested) immediately.

Signed-off-by: NeilBrown <neilb@suse.com>
Acked-by: NeilBrown <neilb@suse.com>

---
 net/sunrpc/xprtsock.c |    1 +
 1 file changed, 1 insertion(+)

--- a/net/sunrpc/xprtsock.c
+++ b/net/sunrpc/xprtsock.c
@@ -768,6 +768,7 @@ static void xs_sock_mark_closed(struct r
 	xs_sock_reset_connection_flags(xprt);
 	/* Mark transport as closed and wake up all pending tasks */
 	xprt_disconnect_done(xprt);
+	xprt_force_disconnect(xprt);
 }
 
 /**
